File: tests/oracle/jt_calcrange_1204.cc
/* JT-CALCRANGE-1204 (JUMPTABLE-CALCRANGE-0001): locked Ghidra 12.0.4 fixture.
 *
 * Observes JumpBasic::calcRange (jumptable.cc:1120-1156),
 * JumpBasic::markFoldableGuards (jumptable.cc:1239-1251) and
 * JumpBasic::markModel (jumptable.cc:1254-1267) on synthetic guard CFGs
 * built inside a vehicle Funcdata loaded from the pinned curl binary.
 *
 * calcRange is driven both through the public recoverModel entry (sc1/sc3:
 * jrange size after guard intersection) and directly after analyzeGuards
 * (sc2: constant varnode input).  markModel's skip semantics are observed
 * via PcodeOp::isMark() after markFoldableGuards has cleared the
 * valueMatch==0 guards (their CBRANCH becomes null and their readOp must
 * NOT be marked).
 *
 * Guards use unsigned comparisons only (INT_LESS): the signed pullback and
 * calc_nz_mask residuals are covered by JT-GUARDS-1204 and stay out of
 * this fixture's write-set.
 */
#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "jumptable.hh"

#include <iostream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::map;
using std::ostringstream;
using std::string;
using std::vector;

class TestableJumpBasic : public JumpBasic {
public:
  TestableJumpBasic(JumpTable *jt) : JumpBasic(jt) {}
  using JumpBasic::analyzeGuards;
  using JumpBasic::calcRange;
  using JumpBasic::markModel;
  vector<GuardRecord> &guards(void) { return selectguards; }
};

struct Lab {
  Funcdata *fd;
  AddrSpace *code;
  map<const BlockBasic *, string> blockName;
  map<const Varnode *, string> varName;
  map<const PcodeOp *, string> opName;
  int seq;

  Lab(Funcdata *f) : fd(f), code(f->getArch()->getDefaultCodeSpace()), seq(0) {}

  BlockBasic *blk(const string &name)
  {
    BlockGraph &graph = const_cast<BlockGraph &>(fd->getBasicBlocks());
    BlockBasic *b = graph.newBlockBasic(fd);
    blockName[b] = name;
    return b;
  }
  Varnode *var(const string &name,int4 size)
  {
    Varnode *vn = fd->newUnique(size);
    varName[vn] = name;
    return vn;
  }
  Varnode *cnst(const string &name,int4 size,uintb val)
  {
    Varnode *vn = fd->newConstant(size,val);
    varName[vn] = name;
    return vn;
  }
  Varnode *coderef(const string &name,uintb off)
  {
    Varnode *vn = fd->newCodeRef(Address(code,off));
    varName[vn] = name;
    return vn;
  }
  PcodeOp *op(const string &name,BlockBasic *block,OpCode opc,int4 inputs)
  {
    PcodeOp *o = fd->newOp(inputs,Address(code,0x500000 + (seq++)));
    fd->opSetOpcode(o,opc);
    fd->opInsertEnd(o,block);
    opName[o] = name;
    return o;
  }
  Varnode *out(const string &name,PcodeOp *o,int4 size)
  {
    Varnode *vn = fd->newUniqueOut(size,o);
    varName[vn] = name;
    return vn;
  }
  void in(PcodeOp *o,Varnode *vn,int4 slot) { fd->opSetInput(o,vn,slot); }
  void setInput(Varnode *vn) { fd->setInputVarnode(vn); }
  void edge(BlockBasic *a,BlockBasic *b)
  {
    BlockGraph &graph = const_cast<BlockGraph &>(fd->getBasicBlocks());
    graph.addEdge(a,b);
  }
  string vnName(const Varnode *vn) const
  {
    if (vn == (const Varnode *)0) return string("null");
    map<const Varnode *,string>::const_iterator iter = varName.find(vn);
    return (iter == varName.end()) ? string("?") : (*iter).second;
  }
  string opNameOf(const PcodeOp *o) const
  {
    if (o == (const PcodeOp *)0) return string("null");
    map<const PcodeOp *,string>::const_iterator iter = opName.find(o);
    return (iter == opName.end()) ? string("?") : (*iter).second;
  }
  string rangeRaw(const CircleRange &rng) const
  {
    ostringstream s;
    s << rng.getMin() << '/' << rng.getEnd() << '/' << rng.getMask()
      << '/' << rng.getStep() << '/' << (rng.isEmpty() ? 1 : 0);
    return s.str();
  }
};

// SC1: two chained unsigned range guards (x<8 and x<10) over a 4-byte
// switch variable fed straight into BRANCHIND.  recoverModel drives
// findDeterminingVarnodes + analyzeGuards + findSmallestNormal +
// markFoldableGuards.  calcRange is then probed on: the switch variable
// (guard intersection written back: [0,8) size 8), the boolean comparison
// outputs (isBoolOutput branch: [0,2) restricted by the bool guard to {1}),
// and an unrelated input (no matching guard: full range truncated to the
// positive half).  The jrange observation locks the resulting table size.
void scenarioRangeGuard(Lab &lab)
{
  Varnode *x = lab.var("x",4);
  lab.setInput(x);
  Varnode *u = lab.var("u",4);
  lab.setInput(u);
  BlockBasic *E = lab.blk("E");
  BlockBasic *B = lab.blk("B");
  BlockBasic *S = lab.blk("S");
  BlockBasic *D = lab.blk("D");
  BlockBasic *D2 = lab.blk("D2");
  PcodeOp *cmp1 = lab.op("E.cmp1",E,CPUI_INT_LESS,2);
  Varnode *b0 = lab.out("b0",cmp1,1);
  lab.in(cmp1,x,0);
  lab.in(cmp1,lab.cnst("c10",4,10),1);
  PcodeOp *cbE = lab.op("E.cb",E,CPUI_CBRANCH,2);
  lab.in(cbE,lab.coderef("refE",0x6000),0);
  lab.in(cbE,b0,1);
  PcodeOp *cmp2 = lab.op("B.cmp2",B,CPUI_INT_LESS,2);
  Varnode *b1 = lab.out("b1",cmp2,1);
  lab.in(cmp2,x,0);
  lab.in(cmp2,lab.cnst("c8",4,8),1);
  PcodeOp *cbB = lab.op("B.cb",B,CPUI_CBRANCH,2);
  lab.in(cbB,lab.coderef("refB",0x6010),0);
  lab.in(cbB,b1,1);
  PcodeOp *biS = lab.op("S.bi",S,CPUI_BRANCHIND,1);
  lab.in(biS,x,0);
  lab.edge(E,D);			// out(0): default path
  lab.edge(E,B);			// out(1): switch path
  lab.edge(B,D2);			// out(0): default path
  lab.edge(B,S);			// out(1): switch path

  JumpTable jt(lab.fd->getArch());
  jt.setIndirectOp(biS);
  TestableJumpBasic basic(&jt);
  bool ok = basic.recoverModel(lab.fd,biS,0,500);

  vector<GuardRecord> &guards = basic.guards();
  ostringstream s;
  s << "sc1_range_guard|ok=" << (ok ? 1 : 0);
  s << "|count=" << guards.size();
  for(size_t i=0;i<guards.size();++i) {
    const GuardRecord &g = guards[i];
    s << "|g" << i << ":cb=" << lab.opNameOf(g.getBranch())
      << ",ro=" << lab.opNameOf(g.getReadOp())
      << ",ip=" << g.getPath()
      << ",rng=" << lab.rangeRaw(g.getRange())
      << ",unr=" << (g.isUnrolled() ? 1 : 0);
  }
  CircleRange rng;
  basic.calcRange(x,rng);
  s << "|cr_x=" << lab.rangeRaw(rng);
  basic.calcRange(b0,rng);
  s << "|cr_b0=" << lab.rangeRaw(rng);
  basic.calcRange(b1,rng);
  s << "|cr_b1=" << lab.rangeRaw(rng);
  basic.calcRange(u,rng);
  s << "|cr_u=" << lab.rangeRaw(rng);
  const JumpValuesRange *vr = basic.getValueRange();
  s << "|jsz=" << vr->getSize()
    << "|jsvn=" << lab.vnName(vr->getStartVarnode())
    << "|jsop=" << lab.opNameOf(vr->getStartOp());
  std::cout << s.str() << '\n';
}

// SC2: constant varnode as the CBRANCH boolean (size-1 constant 1 with
// toswitchval=true).  The first GuardRecord's vn is the constant itself,
// so calcRange must NOT early-return on constants: the guard range {1} is
// intersected into single(1) (identity here; the disjoint/empty case needs
// the full CircleRange::intersect port, RANGE-0001).  A 4-byte constant
// with no matching guard keeps its single range and skips the positive
// truncation (size 1).
void scenarioConstantInput(Lab &lab)
{
  BlockBasic *G = lab.blk("G");
  BlockBasic *S = lab.blk("S");
  BlockBasic *D = lab.blk("D");
  Varnode *k1 = lab.cnst("k1",1,1);
  PcodeOp *cbG = lab.op("G.cb",G,CPUI_CBRANCH,2);
  lab.in(cbG,lab.coderef("refG",0x6200),0);
  lab.in(cbG,k1,1);
  PcodeOp *biS = lab.op("S.bi",S,CPUI_BRANCHIND,1);
  lab.in(biS,lab.var("w",4),0);
  lab.edge(G,D);			// out(0): default path
  lab.edge(G,S);			// out(1): switch path

  JumpTable jt(lab.fd->getArch());
  jt.setIndirectOp(biS);
  TestableJumpBasic basic(&jt);
  basic.analyzeGuards(S,-1);

  vector<GuardRecord> &guards = basic.guards();
  ostringstream s;
  s << "sc2_constant_input";
  s << "|count=" << guards.size();
  for(size_t i=0;i<guards.size();++i) {
    const GuardRecord &g = guards[i];
    s << "|g" << i << ":cb=" << lab.opNameOf(g.getBranch())
      << ",ro=" << lab.opNameOf(g.getReadOp())
      << ",ip=" << g.getPath()
      << ",rng=" << lab.rangeRaw(g.getRange())
      << ",unr=" << (g.isUnrolled() ? 1 : 0);
  }
  CircleRange rng;
  basic.calcRange(k1,rng);
  s << "|cr_k1=" << lab.rangeRaw(rng);
  Varnode *k2 = lab.cnst("k2",4,0x90000000);
  basic.calcRange(k2,rng);
  s << "|cr_k2=" << lab.rangeRaw(rng);
  std::cout << s.str() << '\n';
}

// SC3: markModel skip semantics.  Same chained-guard shape as SC1; after
// recoverModel the two boolean guards are cleared by markFoldableGuards
// (valueMatch==0 against the selected switch variable x).  markModel(true)
// must mark pathMeld ops and surviving guards' readOps but SKIP the cleared
// guards entirely (getBranch()==null continues before readOp).  markModel
// (false) then clears every mark again.
void scenarioMarkModelSkip(Lab &lab)
{
  Varnode *x = lab.var("x",4);
  lab.setInput(x);
  BlockBasic *E = lab.blk("E");
  BlockBasic *B = lab.blk("B");
  BlockBasic *S = lab.blk("S");
  BlockBasic *D = lab.blk("D");
  BlockBasic *D2 = lab.blk("D2");
  PcodeOp *cmp1 = lab.op("E.cmp1",E,CPUI_INT_LESS,2);
  Varnode *b0 = lab.out("b0",cmp1,1);
  lab.in(cmp1,x,0);
  lab.in(cmp1,lab.cnst("c10",4,10),1);
  PcodeOp *cbE = lab.op("E.cb",E,CPUI_CBRANCH,2);
  lab.in(cbE,lab.coderef("refE",0x6300),0);
  lab.in(cbE,b0,1);
  PcodeOp *cmp2 = lab.op("B.cmp2",B,CPUI_INT_LESS,2);
  Varnode *b1 = lab.out("b1",cmp2,1);
  lab.in(cmp2,x,0);
  lab.in(cmp2,lab.cnst("c8",4,8),1);
  PcodeOp *cbB = lab.op("B.cb",B,CPUI_CBRANCH,2);
  lab.in(cbB,lab.coderef("refB",0x6310),0);
  lab.in(cbB,b1,1);
  PcodeOp *biS = lab.op("S.bi",S,CPUI_BRANCHIND,1);
  lab.in(biS,x,0);
  lab.edge(E,D);
  lab.edge(E,B);
  lab.edge(B,D2);
  lab.edge(B,S);

  JumpTable jt(lab.fd->getArch());
  jt.setIndirectOp(biS);
  TestableJumpBasic basic(&jt);
  bool ok = basic.recoverModel(lab.fd,biS,0,500);

  vector<GuardRecord> &guards = basic.guards();
  ostringstream s;
  s << "sc3_markmodel_skip|ok=" << (ok ? 1 : 0);
  for(size_t i=0;i<guards.size();++i)
    s << "|gn" << i << '=' << (guards[i].getBranch() == (PcodeOp *)0 ? 1 : 0);
  basic.markModel(true);
  s << "|on_Sbi=" << (biS->isMark() ? 1 : 0)
    << "|on_Bcmp2=" << (cmp2->isMark() ? 1 : 0)
    << "|on_Ecmp1=" << (cmp1->isMark() ? 1 : 0)
    << "|on_Bcb=" << (cbB->isMark() ? 1 : 0)
    << "|on_Ecb=" << (cbE->isMark() ? 1 : 0);
  basic.markModel(false);
  s << "|off_Sbi=" << (biS->isMark() ? 1 : 0)
    << "|off_Bcmp2=" << (cmp2->isMark() ? 1 : 0)
    << "|off_Ecmp1=" << (cmp1->isMark() ? 1 : 0)
    << "|off_Bcb=" << (cbB->isMark() ? 1 : 0)
    << "|off_Ecb=" << (cbE->isMark() ? 1 : 0);
  std::cout << s.str() << '\n';
}

void run(const string &specDirectory,const string &binary)
{
  vector<string> specPaths(1,specDirectory);
  startDecompilerLibrary(specPaths);
  {
    ostringstream messages;
    BfdArchitecture architecture(binary,"default",&messages);
    DocumentStorage store;
    architecture.init(store);
    architecture.readLoaderSymbols("::");
    Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
    if (fd == (Funcdata *)0) throw std::runtime_error("function not found: GetStr");
    Lab lab(fd);
    scenarioRangeGuard(lab);
    scenarioConstantInput(lab);
    scenarioMarkModelSkip(lab);
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(int argc,char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: jt_calcrange_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    run(argv[1],argv[2]);
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/jt_calcrange_1204.rs
//! JT-CALCRANGE-1204 (JUMPTABLE-CALCRANGE-0001) Rugra comparand.
//!
//! Mirrors tests/oracle/jt_calcrange_1204.cc: observes
//! JumpBasic::calc_range (jumptable.cc:1120-1156),
//! JumpBasic::mark_foldable_guards (jumptable.cc:1239-1251) and
//! JumpBasic::mark_model (jumptable.cc:1254-1267) on synthetic guard CFGs.
//! Output format is byte-identical to the C++ fixture.

use rugra::address::Address;
use rugra::funcdata::Funcdata;
use rugra::jumptable::{JumpBasic, JumpModel, JumpTable};
use rugra::opcodes::OpCode;
use rugra::rangeutil::CircleRange;
use rugra::varnode::Varnode;
use std::collections::HashMap;
use std::fmt::Write as _;
use std::sync::{Arc, RwLock};

type Block = Arc<RwLock<dyn rugra::block::FlowBlock + Send + Sync>>;
type Var = Arc<RwLock<Varnode>>;
type Op = Arc<RwLock<rugra::op::PcodeOp>>;

/// Mirrors Ghidra's PcodeOp::mark bit as tracked by Rugra's jumptable
/// module (MARK_FLAG in src/jumptable.rs).
const MARK_FLAG: u32 = 1;

struct Lab {
    fd: Funcdata,
    var_name: HashMap<usize, String>,
    op_name: HashMap<usize, String>,
    seq: i32,
}

impl Lab {
    fn new() -> Self {
        Self {
            fd: Funcdata::new("GetStr", Address::new(0x36d0), 0),
            var_name: HashMap::new(),
            op_name: HashMap::new(),
            seq: 0,
        }
    }
    fn blk(&mut self, name: &str) -> (Block, String) {
        let b = self.fd.create_new_block();
        let _ = name;
        (b.clone(), name.to_string())
    }
    fn var(&mut self, name: &str, size: usize) -> Var {
        let vn = self.fd.new_unique(size);
        self.var_name.insert(Arc::as_ptr(&vn) as usize, name.to_string());
        vn
    }
    fn cnst(&mut self, name: &str, size: usize, val: u64) -> Var {
        let vn = self.fd.new_constant(size, val);
        self.var_name.insert(Arc::as_ptr(&vn) as usize, name.to_string());
        vn
    }
    fn coderef(&mut self, name: &str, off: u64) -> Var {
        let vn = self.fd.new_code_ref(Address::new(off));
        self.var_name.insert(Arc::as_ptr(&vn) as usize, name.to_string());
        vn
    }
    fn op(&mut self, name: &str, block: &Block, opc: OpCode, inputs: usize) -> Op {
        let op = self.fd.new_op(inputs, Address::new(0x500000 + self.seq as u64));
        self.seq += 1;
        self.fd.op_set_opcode(&op, opc);
        self.fd.op_insert_end(&op, block);
        let arc = op.0.clone();
        self.op_name.insert(Arc::as_ptr(&arc) as usize, name.to_string());
        arc
    }
    fn out(&mut self, name: &str, op: &Op, size: usize) -> Var {
        let opref = rugra::op::PcodeOpRef(op.clone());
        let vn = self.fd.new_unique_out(size, &opref);
        self.var_name.insert(Arc::as_ptr(&vn) as usize, name.to_string());
        vn
    }
    fn input(&mut self, op: &Op, vn: &Var, slot: usize) {
        let opref = rugra::op::PcodeOpRef(op.clone());
        self.fd.op_set_input(&opref, vn.clone(), slot);
    }
    fn set_input(&mut self, vn: &Var) {
        let _ = self.fd.set_input_varnode(vn.clone());
    }
    fn edge(&mut self, a: &Block, b: &Block) {
        self.fd.bblocks.add_edge(a.clone(), b.clone());
    }
    fn vn_name(&self, vn: Option<&Var>) -> String {
        match vn {
            None => "null".to_string(),
            Some(vn) => self
                .var_name
                .get(&(Arc::as_ptr(vn) as usize))
                .cloned()
                .unwrap_or_else(|| "?".to_string()),
        }
    }
    fn op_name_of(&self, op: &Option<Op>) -> String {
        match op {
            None => "null".to_string(),
            Some(op) => self
                .op_name
                .get(&(Arc::as_ptr(op) as usize))
                .cloned()
                .unwrap_or_else(|| "?".to_string()),
        }
    }
    fn range_raw(rng: &CircleRange) -> String {
        format!(
            "{}/{}/{}/{}/{}",
            rng.get_left(),
            rng.get_right(),
            rng.get_mask(),
            rng.get_step(),
            i32::from(rng.is_empty())
        )
    }
    fn is_marked(op: &Op) -> i32 {
        i32::from((op.read().unwrap().addlflags & MARK_FLAG) != 0)
    }
}

fn dump_guards(lab: &Lab, s: &mut String, guards: &[rugra::jumptable::GuardRecord]) {
    let _ = write!(s, "|count={}", guards.len());
    for (i, g) in guards.iter().enumerate() {
        let _ = write!(
            s,
            "|g{i}:cb={},ro={},ip={},rng={},unr={}",
            lab.op_name_of(&g.get_branch()),
            lab.op_name_of(&g.get_read_op()),
            g.get_path(),
            Lab::range_raw(g.get_range()),
            i32::from(g.is_unrolled())
        );
    }
}

// SC1: two chained unsigned range guards (x<8 and x<10) over a 4-byte
// switch variable fed straight into BRANCHIND.
fn scenario_range_guard(lab: &mut Lab) {
    let x = lab.var("x", 4);
    lab.set_input(&x);
    let u = lab.var("u", 4);
    lab.set_input(&u);
    let (e, _) = lab.blk("E");
    let (b, _) = lab.blk("B");
    let (s, _) = lab.blk("S");
    let (d, _) = lab.blk("D");
    let (d2, _) = lab.blk("D2");
    let cmp1 = lab.op("E.cmp1", &e, OpCode::CPUI_INT_LESS, 2);
    let b0 = lab.out("b0", &cmp1, 1);
    lab.input(&cmp1, &x, 0);
    let c10 = lab.cnst("c10", 4, 10);
    lab.input(&cmp1, &c10, 1);
    let cb_e = lab.op("E.cb", &e, OpCode::CPUI_CBRANCH, 2);
    let ref_e = lab.coderef("refE", 0x6000);
    lab.input(&cb_e, &ref_e, 0);
    lab.input(&cb_e, &b0, 1);
    let cmp2 = lab.op("B.cmp2", &b, OpCode::CPUI_INT_LESS, 2);
    let b1 = lab.out("b1", &cmp2, 1);
    lab.input(&cmp2, &x, 0);
    let c8 = lab.cnst("c8", 4, 8);
    lab.input(&cmp2, &c8, 1);
    let cb_b = lab.op("B.cb", &b, OpCode::CPUI_CBRANCH, 2);
    let ref_b = lab.coderef("refB", 0x6010);
    lab.input(&cb_b, &ref_b, 0);
    lab.input(&cb_b, &b1, 1);
    let bi_s = lab.op("S.bi", &s, OpCode::CPUI_BRANCHIND, 1);
    lab.input(&bi_s, &x, 0);
    lab.edge(&e, &d); // out(0): default path
    lab.edge(&e, &b); // out(1): switch path
    lab.edge(&b, &d2); // out(0): default path
    lab.edge(&b, &s); // out(1): switch path

    let jt = Arc::new(RwLock::new(JumpTable::new(Address::new(0))));
    jt.write().unwrap().set_indirect_op(bi_s.clone());
    let mut basic = JumpBasic::new(jt.clone());
    // recover_model now returns Result<bool, JumpTableRecoveryError> (the
    // LowlevelError channel Ghidra raises as an exception). In these scenarios
    // the channel must stay silent; a raised error mirrors the C++ uncaught
    // LowlevelError abort, so propagate it loudly instead of printing ok=0.
    let ok = basic
        .recover_model(&lab.fd, &bi_s, 0, 500)
        .expect("sc1 recover_model error channel must stay silent");

    let mut s = String::new();
    let _ = write!(s, "sc1_range_guard|ok={}", i32::from(ok));
    dump_guards(lab, &mut s, &basic.selectguards);
    let mut rng = CircleRange::empty();
    basic.calc_range(&x, &mut rng);
    let _ = write!(s, "|cr_x={}", Lab::range_raw(&rng));
    basic.calc_range(&b0, &mut rng);
    let _ = write!(s, "|cr_b0={}", Lab::range_raw(&rng));
    basic.calc_range(&b1, &mut rng);
    let _ = write!(s, "|cr_b1={}", Lab::range_raw(&rng));
    basic.calc_range(&u, &mut rng);
    let _ = write!(s, "|cr_u={}", Lab::range_raw(&rng));
    let vr = basic.jrange.as_ref().expect("jrange allocated by recover_model");
    let _ = write!(
        s,
        "|jsz={}|jsvn={}|jsop={}",
        vr.get_size(),
        lab.vn_name(vr.get_start_varnode().as_ref()),
        lab.op_name_of(&vr.get_start_op())
    );
    println!("{s}");
}

// SC2: constant varnode as the CBRANCH boolean.
fn scenario_constant_input(lab: &mut Lab) {
    let (g, _) = lab.blk("G");
    let (s2, _) = lab.blk("S");
    let (d3, _) = lab.blk("D");
    let k1 = lab.cnst("k1", 1, 1);
    let cb_g = lab.op("G.cb", &g, OpCode::CPUI_CBRANCH, 2);
    let ref_g = lab.coderef("refG", 0x6200);
    lab.input(&cb_g, &ref_g, 0);
    lab.input(&cb_g, &k1, 1);
    let bi_s = lab.op("S.bi", &s2, OpCode::CPUI_BRANCHIND, 1);
    let w = lab.var("w", 4);
    lab.input(&bi_s, &w, 0);
    lab.edge(&g, &d3); // out(0): default path
    lab.edge(&g, &s2); // out(1): switch path

    let jt = Arc::new(RwLock::new(JumpTable::new(Address::new(0))));
    jt.write().unwrap().set_indirect_op(bi_s.clone());
    let mut basic = JumpBasic::new(jt.clone());
    basic.analyze_guards(&s2, -1);

    let mut s = String::from("sc2_constant_input");
    dump_guards(lab, &mut s, &basic.selectguards);
    let mut rng = CircleRange::empty();
    basic.calc_range(&k1, &mut rng);
    let _ = write!(s, "|cr_k1={}", Lab::range_raw(&rng));
    let k2 = lab.cnst("k2", 4, 0x90000000);
    basic.calc_range(&k2, &mut rng);
    let _ = write!(s, "|cr_k2={}", Lab::range_raw(&rng));
    println!("{s}");
}

// SC3: markModel skip semantics.
fn scenario_mark_model_skip(lab: &mut Lab) {
    let x = lab.var("x", 4);
    lab.set_input(&x);
    let (e, _) = lab.blk("E");
    let (b, _) = lab.blk("B");
    let (s3, _) = lab.blk("S");
    let (d, _) = lab.blk("D");
    let (d2, _) = lab.blk("D2");
    let cmp1 = lab.op("E.cmp1", &e, OpCode::CPUI_INT_LESS, 2);
    let b0 = lab.out("b0", &cmp1, 1);
    lab.input(&cmp1, &x, 0);
    let c10 = lab.cnst("c10", 4, 10);
    lab.input(&cmp1, &c10, 1);
    let cb_e = lab.op("E.cb", &e, OpCode::CPUI_CBRANCH, 2);
    let ref_e = lab.coderef("refE", 0x6300);
    lab.input(&cb_e, &ref_e, 0);
    lab.input(&cb_e, &b0, 1);
    let cmp2 = lab.op("B.cmp2", &b, OpCode::CPUI_INT_LESS, 2);
    let b1 = lab.out("b1", &cmp2, 1);
    lab.input(&cmp2, &x, 0);
    let c8 = lab.cnst("c8", 4, 8);
    lab.input(&cmp2, &c8, 1);
    let cb_b = lab.op("B.cb", &b, OpCode::CPUI_CBRANCH, 2);
    let ref_b = lab.coderef("refB", 0x6310);
    lab.input(&cb_b, &ref_b, 0);
    lab.input(&cb_b, &b1, 1);
    let bi_s = lab.op("S.bi", &s3, OpCode::CPUI_BRANCHIND, 1);
    lab.input(&bi_s, &x, 0);
    lab.edge(&e, &d);
    lab.edge(&e, &b);
    lab.edge(&b, &d2);
    lab.edge(&b, &s3);

    let jt = Arc::new(RwLock::new(JumpTable::new(Address::new(0))));
    jt.write().unwrap().set_indirect_op(bi_s.clone());
    let mut basic = JumpBasic::new(jt.clone());
    // Same Result adaptation as sc1: the error channel must stay silent.
    let ok = basic
        .recover_model(&lab.fd, &bi_s, 0, 500)
        .expect("sc3 recover_model error channel must stay silent");

    let mut s = String::new();
    let _ = write!(s, "sc3_markmodel_skip|ok={}", i32::from(ok));
    for (i, g) in basic.selectguards.iter().enumerate() {
        let _ = write!(s, "|gn{i}={}", i32::from(g.get_branch().is_none()));
    }
    basic.mark_model(true);
    let _ = write!(
        s,
        "|on_Sbi={}|on_Bcmp2={}|on_Ecmp1={}|on_Bcb={}|on_Ecb={}",
        Lab::is_marked(&bi_s),
        Lab::is_marked(&cmp2),
        Lab::is_marked(&cmp1),
        Lab::is_marked(&cb_b),
        Lab::is_marked(&cb_e)
    );
    basic.mark_model(false);
    let _ = write!(
        s,
        "|off_Sbi={}|off_Bcmp2={}|off_Ecmp1={}|off_Bcb={}|off_Ecb={}",
        Lab::is_marked(&bi_s),
        Lab::is_marked(&cmp2),
        Lab::is_marked(&cmp1),
        Lab::is_marked(&cb_b),
        Lab::is_marked(&cb_e)
    );
    println!("{s}");
}

fn main() {
    let mut lab = Lab::new();
    scenario_range_guard(&mut lab);
    scenario_constant_input(&mut lab);
    scenario_mark_model_skip(&mut lab);
}
